src/protojure/pedestal/interceptors/grpc_web.clj

```diff
@@ -4,24 +4,65 @@
 
 (ns protojure.pedestal.interceptors.grpc-web
   "A [Pedestal](http://pedestal.io/) [interceptor](http://pedestal.io/reference/interceptors) for the [GRPC-WEB](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md) protocol"
-  (:require [io.pedestal.interceptor :refer [->Interceptor]])
-  (:import (org.apache.commons.codec.binary Base64InputStream))
+  (:require [io.pedestal.interceptor :refer [->Interceptor]]
+            [clojure.core.async :as async]
+            [clojure.data])
   (:refer-clojure :exclude [proxy]))
 
 (set! *warn-on-reflection* true)
 
+(defn read-n [from-chan n]
+  "Convenience method for consuming n [n] or less values from a channel [from-chan]"
+  (async/go-loop [res []]
+    (let [v (async/<! from-chan)]
+      (if (nil? v)
+        [true res]
+        (let [new-res (conj res v)]
+          (if (= n (count new-res))
+            [false new-res]
+            (recur new-res)))))))
+
 (defn- decode-body
-  [{:keys [body] :as request}]
-  (assoc request :body (Base64InputStream. body)))
+  "Consumes 4 base64 encoded bytes from the body, and writes to a new channel that replaces the prior :body-ch on the
+  request. Note that any consumption from the :body inputstream that is not followed by a write to the :body-ch key
+  results in that data being lost to the grpc interceptor :grpc-params/:body-ch key."
+  [{:keys [body-ch] :as request}]
+  (let [dec-ch (async/chan 4056)
+        decoder (java.util.Base64/getDecoder)]
+    (async/go-loop [[final encoded] (async/<! (read-n body-ch 4))]
+      (if (and (empty? encoded) final)
+        (async/close! dec-ch)
+        (do
+          (doseq [b (.decode decoder (byte-array encoded))]
+            (async/>! dec-ch b))
+          (recur (async/<! (read-n body-ch 4))))))
+    (assoc request :body-ch dec-ch)))
+
+(defn- encode-body
+  "Consumes bytes from the response body channel and base64 encodes the payload"
+  [{{:keys [body] :as response} :response :as ctx}]
+  (let [encoder (java.util.Base64/getEncoder)
+        out-ch (async/chan 4056)]
+    (async/go-loop [s (async/<! body)]
```

Comment:
What about cases where the body is not a channel.  I might be misunderstanding, but it seems like this would run simply with any request that has Accepts: grpc-web-text, which may or may not have found a grpc-interceptor route.  If that is the case, you might need to have something like a multi-method on the body-type.

Reply:
I'm fairly certain the regular grpc interceptor is going make it a channel for the lpm/encode, and this must run AFTER lpm/encode: https://github.com/protojure/lib/blob/master/src/protojure/pedestal/interceptors/grpc.clj#L105

Reply:
Agree its unlikely, and we can iterate on it.  To be defensive, perhaps we should just generate a clean error if its not a channel?

Reply:
This is done. We throw and ultimately return a grpc-status 13.